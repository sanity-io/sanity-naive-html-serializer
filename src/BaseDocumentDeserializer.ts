import schema from 'part:@sanity/base/schema'
import Schema from '@sanity/schema'
import blockTools from '@sanity/block-tools'
import {
  customDeserializers,
  customBlockDeserializers,
} from './BaseSerializationConfig'
import { ObjectField, ObjectSchemaType, BlockSchemaType } from '@sanity/types'
import { Deserializer } from './types'

const noSchemaWarning = (obj: Element) =>
  `WARNING: Unfortunately the deserializer may have issues with this field or object: ${obj.className}.
  If it's a specific type, you may need to declare  at the top level, or write a custom deserializer.`

const defaultSchema = Schema.compile({
  name: 'default',
  types: [
    {
      type: 'object',
      name: 'default',
      fields: [
        {
          name: 'block',
          type: 'array',
          of: [{ type: 'block' }],
        },
      ],
    },
  ],
})

const blockContentType = defaultSchema
  .get('default')
  .fields.find((field: ObjectField) => field.name === 'block').type

//helper to handle messy input -- take advantage
//of blockTools' sanitizing behavior for single strings
const preprocess = (html: string) => {
  const intermediateBlocks = blockTools.htmlToBlocks(
    `<p>${html}</p>`,
    blockContentType
  )
  if (!intermediateBlocks.length) {
    throw new Error(`Error parsing string '${html}'`)
  }
  return intermediateBlocks[0].children[0].text
}

export const deserializeDocument = (
  serializedDoc: string,
  deserializers: Record<string, any> = customDeserializers,
  blockDeserializers = customBlockDeserializers
) => {
  const metadata: Record<string, any> = {}
  const head = new DOMParser().parseFromString(serializedDoc, 'text/html').head

  Array.from(head.children).forEach(metaTag => {
    const validTags = ['_id', '_rev', '_type']
    const metaName = metaTag.getAttribute('name')
    if (metaName && validTags.includes(metaName)) {
      metadata[metaName] = metaTag.getAttribute('content')
    }
  })

  const content: Record<string, any> = deserializeHTML(
    serializedDoc,
    schema.get(metadata._type),
    deserializers,
    blockDeserializers
  )

  return {
    ...metadata,
    ...content,
  }
}

const deserializeHTML = (
  html: string,
  target: ObjectSchemaType | BlockSchemaType,
  deserializers: Record<string, any>,
  blockDeserializers: Array<any>
): Record<string, any> | any[] => {
  //parent node is always div with classname of field -- get its children
  let HTMLnode = new DOMParser().parseFromString(html, 'text/html').body
    .children[0]
  if (!HTMLnode) {
    return {}
  }
  //@ts-ignore
  let output

  //catch embedded object as a field -- not necessarily foolproof
  if (
    HTMLnode.children.length === 1 &&
    target &&
    target.type &&
    HTMLnode.children[0].className === target.type.name
  ) {
    HTMLnode = HTMLnode.children[0]
  }

  //prioritize custom deserialization
  if (
    target &&
    target.type &&
    target.type.name &&
    Object.keys(deserializers.types).includes(target.type.name)
  ) {
    const deserialize = deserializers.types[target.type.name]
    output = deserialize(HTMLnode)
  }

  //begin recursively deserializing
  //every child node is either a field in an object
  //or a child object in an array
  else {
    const children = Array.from(HTMLnode.children)
    output = target && target.type && target.type.hasOwnProperty('of') ? [] : {}

    children.forEach(child => {
      let deserializedObj
      //prioritize custom deserialization
      if (Object.keys(deserializers.types).includes(child.className)) {
        const deserialize = deserializers.types[child.className]
        deserializedObj = deserialize(child)
      }

      //flat string, it's an unrich field
      else if (child.tagName.toLowerCase() === 'span') {
        deserializedObj = preprocess(child.innerHTML)
      }

      //has specific class name, so it's either a field or obj
      else if (child.className) {
        let objType
        //wrapper is fieldname
        if (target && target.fields) {
          objType = target.fields.find(field => field.name === child.className)
        }

        if (!objType && schema.get(child.className)) {
          objType = schema.get(child.className)
        }

        //@ts-ignore
        if (!objType && target && target.type && target.type.fields) {
          //@ts-ignore
          objType = target.type.fields.find(
            (field: ObjectField) => field.name === child.className
          )
        }

        if (!objType) {
          console.debug(noSchemaWarning(child))
          objType = blockContentType
        }

        try {
          deserializedObj = deserializeHTML(
            child.outerHTML,
            objType,
            deserializers,
            blockDeserializers
          )
        } catch (err) {
          console.debug(err)
          try {
            deserializedObj = deserializeHTML(
              child.outerHTML,
              schema.get('object'),
              deserializers,
              blockDeserializers
            )
          } catch (err) {
            console.debug(
              `Tried to deserialize block of type ${child.className} but failed! Received error ${err}`
            )
          }
        }

        // @ts-ignore
        if (Array.isArray(output) && deserializedObj) {
          deserializedObj._type = child.className
          deserializedObj._key = child.id
        }
      } else {
        deserializedObj = blockTools.htmlToBlocks(
          child.outerHTML,
          blockContentType,
          { rules: blockDeserializers }
        )[0]
        deserializedObj._key = child.id
      }

      // @ts-ignore
      if (Array.isArray(output)) {
        output.push(deserializedObj)
      } else {
        // @ts-ignore
        output[child.className] = deserializedObj
      }
    })
  }

  return output
}

export const BaseDocumentDeserializer: Deserializer = {
  deserializeDocument,
  deserializeHTML,
}
