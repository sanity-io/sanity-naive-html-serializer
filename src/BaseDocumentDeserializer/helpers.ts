import {htmlToBlocks} from '@sanity/block-tools'
import {Schema} from '@sanity/schema'
import {ObjectField, PortableTextSpan, PortableTextTextBlock} from 'sanity'

const defaultSchema = Schema.compile({
  name: 'default',
  types: [
    {
      type: 'object',
      name: 'default',
      fields: [
        {
          name: 'block',
          type: 'array',
          of: [{type: 'block'}],
        },
      ],
    },
  ],
})

export const blockContentType = defaultSchema
  .get('default')
  .fields.find((field: ObjectField) => field.name === 'block').type

export const noSchemaWarning = (obj: Element): string =>
  `WARNING: Unfortunately the deserializer may have issues with this field or object: ${obj.className}.
  If it's a specific type, you may need to declare  at the top level, or write a custom deserializer.`

//helper to handle messy input -- take advantage
//of blockTools' sanitizing behavior for single strings
export const preprocess = (html: string): string => {
  const intermediateBlocks = htmlToBlocks(
    `<p>${html}</p>`,
    blockContentType
  ) as PortableTextTextBlock<PortableTextSpan>[]
  if (!intermediateBlocks.length) {
    throw new Error(`Error parsing string '${html}'`)
  }
  return intermediateBlocks[0].children[0].text
}
