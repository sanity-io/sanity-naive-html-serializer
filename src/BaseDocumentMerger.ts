import {Merger} from './types'
import {SanityDocument} from 'sanity'
import {extractWithPath, arrayToJSONMatchPath} from '@sanity/mutator'

const reconcileArray = (origArray: any[], translatedArray: any[]): any[] => {
  //arrays of strings don't have keys, so just replace the array and return
  if (translatedArray && translatedArray.some((el) => typeof el === 'string')) {
    return translatedArray
  }

  //deep copy needed for field level patching
  const combined = JSON.parse(JSON.stringify(origArray))

  translatedArray.forEach((block) => {
    if (!block._key) {
      return
    }
    const foundBlockIdx = origArray.findIndex((origBlock) => origBlock._key === block._key)
    if (foundBlockIdx < 0) {
      //eslint-disable-next-line no-console
      console.warn(
        `This block no longer exists on the original document. Was it removed? ${JSON.stringify(
          block
        )}`
      )
    } else if (
      origArray[foundBlockIdx]._type === 'block' ||
      origArray[foundBlockIdx]._type === 'span'
    ) {
      combined[foundBlockIdx] = block
    } else if (Array.isArray(origArray[foundBlockIdx])) {
      combined[foundBlockIdx] = reconcileArray(origArray[foundBlockIdx], block)
    } else {
      //eslint-disable-next-line no-use-before-define -- this is a recursive function
      combined[foundBlockIdx] = reconcileObject(origArray[foundBlockIdx], block)
    }
  })
  return combined
}

const reconcileObject = (
  origObject: Record<string, any>,
  translatedObject: Record<string, any>
): Record<string, any> => {
  if (typeof translatedObject !== 'object' || !Object.keys(translatedObject).length) {
    return origObject
  }

  const updatedObj = JSON.parse(JSON.stringify(origObject))
  Object.entries(translatedObject).forEach(([key, value]) => {
    if (!value || key[0] === '_') {
      return
    }
    if (typeof value === 'string') {
      updatedObj[key] = value
    } else if (Array.isArray(value)) {
      //eslint-disable-next-line @typescript-eslint/no-use-before-define -- this is a recursive function
      updatedObj[key] = reconcileArray(origObject[key] ?? [], value)
    } else {
      updatedObj[key] = reconcileObject(origObject[key] ?? {}, value)
    }
  })
  return updatedObj
}

const fieldLevelMerge = (
  translatedFields: Record<string, any>,
  //should be fetched according to the revision and id of the translated obj above
  baseDoc: SanityDocument,
  localeId: string,
  baseLang: string = 'en'
): Record<string, any> => {
  const merged: Record<string, any> = {}
  const metaKeys = ['_rev', '_id', '_type']
  metaKeys.forEach((metaKey) => {
    if (translatedFields[metaKey]) {
      merged[metaKey] = translatedFields[metaKey]
    }
  })

  //get any field that matches the base language, because it's been translated
  const originPaths = extractWithPath(`..${baseLang}`, translatedFields)
  originPaths.forEach((match) => {
    const origVal = extractWithPath(arrayToJSONMatchPath(match.path), baseDoc)[0].value
    const translatedVal = extractWithPath(arrayToJSONMatchPath(match.path), translatedFields)[0]
      .value
    let valToPatch
    if (typeof translatedVal === 'string') {
      valToPatch = translatedVal
    } else if (Array.isArray(translatedVal) && translatedVal.length) {
      valToPatch = reconcileArray((origVal as Array<any>) ?? [], translatedVal)
    } else if (
      typeof translatedVal === 'object' &&
      Object.keys(translatedVal as Record<string, any>).length
    ) {
      valToPatch = reconcileObject(origVal ?? {}, translatedVal as Record<string, any>)
    }
    const destinationPath = [
      ...match.path.slice(0, match.path.length - 1), //cut off the "en"
      localeId.replace('-', '_'), // replace it with our locale
    ]

    merged[arrayToJSONMatchPath(destinationPath)] = valToPatch
  })

  return merged
}

const documentLevelMerge = (
  translatedFields: Record<string, any>,
  //should be fetched according to the revision and id of the translated obj above
  baseDoc: SanityDocument
): Record<string, any> => {
  return reconcileObject(baseDoc, translatedFields)
}

export const BaseDocumentMerger: Merger = {
  fieldLevelMerge,
  documentLevelMerge,
  reconcileArray,
  reconcileObject,
}
