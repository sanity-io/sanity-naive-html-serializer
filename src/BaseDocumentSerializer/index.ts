import {defaultStopTypes, customSerializers} from '../BaseSerializationConfig'
import {SanityDocument, TypedObject, Schema} from 'sanity'
import {TranslationLevel, SerializerClosure} from '../types'
import {fieldFilter, languageObjectFieldFilter} from './fieldFilters'
import {PortableTextTypeComponent, toHTML} from '@portabletext/to-html'

const META_FIELDS = ['_key', '_type', '_id', '_weak']

export const BaseDocumentSerializer: SerializerClosure = (schemas: Schema) => {
  /*
   * Helper function that allows us to get metadata (like `localize: false`) from schema fields.
   */
  const getSchema = (name: string) => schemas?._original?.types.find((s) => s.name === name) as any

  const serializeObject = (
    obj: TypedObject,
    stopTypes: string[],
    serializers: Record<string, any>
  ) => {
    if (stopTypes.includes(obj._type)) {
      return ''
    }

    //if user has declared a custom serializer, use that
    //instead of this method
    const hasSerializer = serializers.types && Object.keys(serializers.types).includes(obj._type)
    if (hasSerializer) {
      return toHTML([obj], {components: serializers})
    }

    //we don't need to worry about PT types
    if (obj._type === 'span' || obj._type === 'block') {
      return toHTML(obj, {components: serializers})
    }

    //if schema is available, encode values in the order they're declared in the schema,
    //since this will likely be more intuitive for a translator.
    let fieldNames = Object.keys(obj).filter((key) => key !== '_type')
    const schema = getSchema(obj._type)
    if (schema && schema.fields) {
      fieldNames = schema.fields
        .map((field: Record<string, any>) => field.name)
        .filter((schemaKey: string) => Object.keys(obj).includes(schemaKey))
    }

    //account for anonymous inline objects
    if (typeof obj === 'object' && !obj._type) {
      obj._type = ''
    }

    //in some cases, we might recurse through many objects of the same type
    //we should take all methods necessary to ensure state does not persist
    //otherwise we risk using old serialization methods on new items
    const newSerializationMethods: Record<string, PortableTextTypeComponent> = {}
    const tempType = `${obj._type}__temp_type__${Math.random().toString(36).substring(7)}`
    const objToSerialize: TypedObject = {_type: tempType}
    //for our default serialization method, we only need to
    //capture metadata. the rest will be recursively turned into strings.
    META_FIELDS.filter((f) => f !== '_type').forEach((field) => {
      objToSerialize[field] = obj[field]
    })

    let innerHTML = ''

    //if it's a custom object, iterate through its keys to find and serialize translatable content
    fieldNames.forEach((fieldName) => {
      let htmlField = ''

      if (!META_FIELDS.includes(fieldName)) {
        const value = obj[fieldName]
        //strings are either string fields or have recursively been turned
        //into HTML because they were a nested object or array
        if (typeof value === 'string') {
          const htmlRegex = new RegExp(/<("[^"]*"|'[^']*'|[^'">])*>/)
          if (htmlRegex.test(value)) {
            htmlField = value
          } else {
            htmlField = `<span class="${fieldName}">${value}</span>`
          }
        }

        //array fields get filtered and its children serialized
        else if (Array.isArray(value)) {
          //eslint-disable-next-line no-use-before-define -- this is a recursive function
          htmlField = serializeArray(value, fieldName, stopTypes, {
            ...serializers,
            types: {...serializers.types},
          })
        }

        //this is an object in an object, serialize it first
        else {
          const embeddedObject = value as TypedObject
          const embeddedObjectSchema = getSchema(embeddedObject._type)
          let toTranslate = embeddedObject
          if (embeddedObjectSchema && embeddedObjectSchema.fields) {
            toTranslate = fieldFilter(toTranslate, embeddedObjectSchema.fields, stopTypes)
          }
          const objHTML = serializeObject(toTranslate, stopTypes, {
            ...serializers,
            types: {...serializers.types},
          })
          htmlField = `<div class="${fieldName}" data-level="field">${objHTML}</div>`
        }

        innerHTML += htmlField
      }
    })

    if (!innerHTML) {
      return ''
    }

    newSerializationMethods[tempType] = ({value}: {value: TypedObject}) => {
      let div = `<div class="${value._type.split('__temp_type__')[0]}"`
      if (value._key || value._id) {
        div += `id="${value._key ?? value._id}"`
      }

      return [div, ` data-type="object">${innerHTML}</div>`].join('')
    }

    let serializedBlock = ''
    try {
      serializedBlock = toHTML(objToSerialize, {
        components: {
          ...serializers,
          types: {
            ...serializers.types,
            ...newSerializationMethods,
          },
        },
      })
    } catch (err) {
      //eslint-disable-next-line no-console -- this is a warning
      console.warn(
        `Had issues serializing block of type "${obj._type}". Please specify a serialization method for this block in your serialization config. Received error: ${err}`
      )
    }

    return serializedBlock
  }

  const serializeArray = (
    fieldContent: Record<string, any>[],
    fieldName: string,
    stopTypes: string[],
    serializers: Record<string, any>
  ) => {
    //filter for any blocks that user has indicated
    //should not be sent for translation
    const validBlocks = fieldContent.filter((block) => !stopTypes.includes(block._type))

    //take out any fields in these blocks that should
    //not be sent to translation
    const filteredBlocks = validBlocks.map((block) => {
      const schema = getSchema(block._type)
      if (schema && schema.fields) {
        return fieldFilter(block, schema.fields, stopTypes)
      }
      return block
    })

    const output = filteredBlocks.map((obj) => {
      //if object in array is just a string, just return it
      if (typeof obj === 'string') {
        return `<span>${obj}</span>`
      }
      //send to serialization method
      return serializeObject(obj as TypedObject, stopTypes, serializers)
    })

    //encode this with data-level field
    return `<div class="${fieldName}" data-type="array">${output.join('')}</div>`
  }

  /*
   * Main parent function: finds fields to translate, and feeds them to appropriate child serialization
   * methods.
   */
  const serializeDocument = (
    doc: SanityDocument,
    translationLevel: TranslationLevel = 'document',
    baseLang = 'en',
    stopTypes = defaultStopTypes,
    serializers = customSerializers
  ) => {
    const schema = getSchema(doc._type)
    let filteredObj: Record<string, any> = {}

    //field level translations explicitly send over any fields that
    //match the base language, regardless of depth
    if (translationLevel === 'field') {
      filteredObj = languageObjectFieldFilter(doc, baseLang)
    }
    //otherwise, we can refer to the schema and a list of stop types
    //to determine what should not be sent
    else {
      filteredObj = fieldFilter(doc, schema.fields, stopTypes)
    }

    const serializedFields: Record<string, any> = {}

    for (const key in filteredObj) {
      if (filteredObj.hasOwnProperty(key) === false) continue
      const value: Record<string, any> | Array<any> | string = filteredObj[key]

      if (typeof value === 'string') {
        serializedFields[key] = value
      } else if (Array.isArray(value)) {
        serializedFields[key] = serializeArray(value, key, stopTypes, serializers)
      } else if (value && !stopTypes.find((stopType) => stopType == value?._type)) {
        const serialized = serializeObject(value as TypedObject, stopTypes, serializers)
        serializedFields[key] = `<div class="${key}" data-level='field'>${serialized}</div>`
      }
    }

    //create a valid HTML file
    const rawHTMLBody = document.createElement('body')
    rawHTMLBody.innerHTML = serializeObject(serializedFields as TypedObject, stopTypes, serializers)

    const rawHTMLHead = document.createElement('head')
    const metaFields = ['_id', '_type', '_rev']
    //save our metadata as meta tags so we can use them later on
    metaFields.forEach((field) => {
      const metaEl = document.createElement('meta')
      metaEl.setAttribute('name', field)
      metaEl.setAttribute('content', doc[field] as string)
      rawHTMLHead.appendChild(metaEl)
    })
    //encode version so we can use the correct deserialization methods
    const versionMeta = document.createElement('meta')
    versionMeta.setAttribute('name', 'version')
    versionMeta.setAttribute('content', '3')
    rawHTMLHead.appendChild(versionMeta)

    const rawHTML = document.createElement('html')
    rawHTML.appendChild(rawHTMLHead)
    rawHTML.appendChild(rawHTMLBody)

    return {
      name: doc._id,
      content: rawHTML.outerHTML,
    }
  }

  return {
    serializeDocument,
    fieldFilter,
    languageObjectFieldFilter,
    serializeArray,
    serializeObject,
  }
}
