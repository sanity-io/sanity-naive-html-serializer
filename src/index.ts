export {BaseDocumentMerger} from './BaseDocumentMerger'
export {BaseDocumentSerializer} from './BaseDocumentSerializer'
export {BaseDocumentDeserializer} from './BaseDocumentDeserializer'
export {
  defaultStopTypes,
  customSerializers,
  customBlockDeserializers,
} from './BaseSerializationConfig'

export type {SerializedDocument, Serializer, SerializerClosure, Deserializer, Merger} from './types'
