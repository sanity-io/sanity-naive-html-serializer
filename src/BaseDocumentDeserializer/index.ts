import BaseDocumentDeserializer from './BaseDocumentDeserializer'
import LegacyBaseDocumentDeserializer from './LegacyBaseDocumentDeserializer'

export { LegacyBaseDocumentDeserializer, BaseDocumentDeserializer }
