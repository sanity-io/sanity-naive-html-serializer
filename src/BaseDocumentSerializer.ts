import schemas from 'part:@sanity/base/schema'
import blocksToHtml, { h } from '@sanity/block-content-to-html'
import { defaultStopTypes, customSerializers } from './BaseSerializationConfig'
import { ObjectField } from '@sanity/types'
import { findLatestDraft } from './helpers'
import { Serializer } from './types'
import clone from 'just-clone'

/*
 * Helper function that allows us to get metadata (like `localize: false`) from schema fields.
 */
const getSchema = (name: string) =>
  schemas._original.types.find((s: ObjectField) => s.name === name)

/*
 * Main parent function: finds fields to translate, and feeds them to appropriate child serialization
 * methods.
 */
const serializeDocument = async (
  documentId: string,
  translationLevel: string = 'document',
  baseLang: string = 'en',
  stopTypes: string[] = defaultStopTypes,
  serializers: Record<string, any> = customSerializers
) => {
  const doc = await findLatestDraft(documentId)

  let filteredObj: Record<string, any> = {}

  if (translationLevel === 'field') {
    filteredObj = languageObjectFieldFilter(doc, baseLang)
  } else {
    filteredObj = fieldFilter(doc, getSchema(doc._type).fields, stopTypes)
  }

  const serializedFields: Record<string, any> = {}
  for (let key in filteredObj) {
    const value: Record<string, any> | Array<any> = filteredObj[key]
    if (typeof value === 'string') {
      serializedFields[key] = value
    } else if (Array.isArray(value)) {
      serializedFields[key] = serializeArray(value, key, stopTypes, serializers)
    } else {
      serializedFields[key] = serializeObject(
        value,
        key,
        stopTypes,
        serializers
      )
    }
  }

  return {
    name: documentId,
    content: serializeObject(
      serializedFields,
      doc._type,
      stopTypes,
      serializers
    ),
  }
}

/*
 * Helper. If field-level translation pattern used, only sends over
 * content from the base language.
 */
const languageObjectFieldFilter = (
  obj: Record<string, any>,
  baseLang: string
) => {
  const filteredObj: Record<string, any> = {}
  for (let key in obj) {
    const value: any = obj[key]
    if (value.hasOwnProperty(baseLang)) {
      filteredObj[key] = {}
      filteredObj[key][baseLang] = value[baseLang]
    }
  }

  return filteredObj
}

/*
 * Helper. Eliminates stop-types and non-localizable fields.
 */
const fieldFilter = (
  obj: Record<string, any>,
  objFields: ObjectField[],
  stopTypes: string[]
) => {
  const filteredObj: Record<string, any> = {}

  const fieldFilter = (field: Record<string, any>) => {
    if (field.localize === false) {
      return false
    } else if (field.type.name === 'string' || field.type.name === 'text') {
      return true
    } else if (Array.isArray(obj[field.name])) {
      return true
    } else if (!stopTypes.includes(field.type.name)) {
      return true
    }
    return false
  }

  const validFields = [
    '_key',
    '_type',
    '_id',
    ...objFields.filter(fieldFilter).map(field => field.name),
  ]
  validFields.forEach(field => {
    if (obj[field]) {
      filteredObj[field] = obj[field]
    }
  })
  return filteredObj
}

const serializeArray = (
  fieldContent: Record<string, any>[],
  fieldName: string,
  stopTypes: string[],
  serializers: Record<string, any>
) => {
  const validBlocks = fieldContent.filter(
    block => !stopTypes.includes(block._type)
  )

  const filteredBlocks = validBlocks.map(block => {
    const schema = getSchema(block._type)
    if (schema) {
      return fieldFilter(block, schema.fields, stopTypes)
    } else {
      return block
    }
  })

  const output = filteredBlocks.map(obj =>
    //deep copy needed here to avoid serializer rules
    //serializing objects with content that comes before in the loop
    serializeObject(obj, null, stopTypes, serializers)
  )

  return `<div class="${fieldName}">${output.join('')}</div>`
}

const serializeObject = (
  obj: Record<string, any>,
  topFieldName: string | null = null,
  stopTypes: string[],
  serializers: Record<string, any>
) => {
  if (stopTypes.includes(obj._type)) {
    return ''
  }

  const hasSerializer =
    serializers.types && Object.keys(serializers.types).includes(obj._type)
  if (hasSerializer) {
    return blocksToHtml({ blocks: [obj], serializers: serializers })
  }

  const tempSerializers = clone(serializers)

  if (obj._type !== 'span' && obj._type !== 'block') {
    let innerHTML = ''
    Object.entries(obj).forEach(([fieldName, value]) => {
      let htmlField = ''

      if (!['_key', '_type', '_id'].includes(fieldName)) {
        if (typeof value === 'string') {
          const htmlRegex = /^</
          //this field may have been recursively turned into html already.
          htmlField = value.match(htmlRegex)
            ? value
            : `<span class="${fieldName}">${value}</span>`
        } else if (Array.isArray(value)) {
          htmlField = serializeArray(value, fieldName, stopTypes, serializers)
        } else {
          const schema = getSchema(value._type)
          let toTranslate = value
          if (schema) {
            toTranslate = fieldFilter(value, schema.fields, stopTypes)
          }
          const objHTML = serializeObject(
            toTranslate,
            null,
            stopTypes,
            serializers
          )
          htmlField = `<div class=${fieldName}>${objHTML}</div>`
        }
      }
      innerHTML += htmlField
    })

    if (!innerHTML) {
      return ''
    }
    tempSerializers.types[obj._type] = (props: Record<string, any>) => {
      return h('div', {
        className: topFieldName ?? props.node._type,
        id: props.node._key ?? props.node._id,
        innerHTML: innerHTML,
      })
    }
  }

  let serializedBlock = ''
  try {
    serializedBlock = blocksToHtml({
      blocks: [obj],
      serializers: tempSerializers,
    })
  } catch (err) {
    console.debug(
      `Had issues serializing block of type "${obj._type}". Please specify a serialization method for this block in your serialization config. Received error: ${err}`
    )
  }

  return serializedBlock
}

export const BaseDocumentSerializer: Serializer = {
  serializeDocument,
  fieldFilter,
  languageObjectFieldFilter,
  serializeArray,
  serializeObject,
}
