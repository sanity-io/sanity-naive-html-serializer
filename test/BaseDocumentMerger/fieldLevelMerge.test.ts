import clone from 'just-clone'
import {BaseDocumentMerger} from '../../src'
import {getNewFieldLevelDocument, getNewObject} from './utils'
import {getDeserialized} from '../helpers'
const fieldLevelArticle = require('../__fixtures__/fieldLevelArticle')
const nestedLanguageFields = require('../__fixtures__/nestedLanguageFields')

const newDocument = getNewFieldLevelDocument()
const fieldLevelPatches = BaseDocumentMerger.fieldLevelMerge(
  newDocument,
  fieldLevelArticle,
  'es_ES',
  'en'
)
test('Global field level snapshot test', () => {
  expect(fieldLevelPatches).toMatchSnapshot()
})

/*
 * Objects
 */
test('Top-level string / text fields from new document patch to new field, and will be maintained in old field', () => {
  expect(fieldLevelPatches['title.es_ES']).toEqual(newDocument.title.en)
  expect(fieldLevelPatches['snippet.es_ES']).toEqual(newDocument.snippet.en)
  expect(fieldLevelPatches['title.es_ES']).not.toEqual(fieldLevelArticle.title.en)
  expect(fieldLevelPatches['snippet.es_ES']).not.toEqual(fieldLevelArticle.snippet.en)
  expect(fieldLevelPatches['title.en']).toBeUndefined()
  expect(fieldLevelPatches['snippet.en']).toBeUndefined()
})

test('Nested object fields override old object fields', () => {
  expect(fieldLevelPatches['config.es_ES'].title).toEqual(newDocument.config.en.title)
  expect(fieldLevelPatches['config.es_ES'].title).not.toEqual(fieldLevelArticle.config.en.title)
  expect(fieldLevelPatches['config.es_ES'].nestedArrayField[0].children[0].text).toEqual(
    newDocument.config.en.nestedArrayField[0].children[0].text
  )
  expect(fieldLevelPatches['config.es_ES'].nestedArrayField[0].children[0].text).not.toEqual(
    fieldLevelArticle.config.en.nestedArrayField[0].children[0].text
  )
})

test('Nested object merge uses old fields when not present on new object', () => {
  expect(fieldLevelPatches['config.es_ES'].objectAsField.content).toEqual(
    fieldLevelArticle.config.en.objectAsField.content
  )
})

/*
 * Arrays
 */
test('Arrays will use new objects when they exist', () => {
  expect(fieldLevelPatches['content.es_ES'][0].children[0].text).toEqual(
    newDocument.content.en[0].children[0].text
  )
  expect(fieldLevelPatches['content.es_ES'][0].children[0].text).not.toEqual(
    fieldLevelArticle.content.en[0].children[0].text
  )
})

test('Arrays will use old blocks if they do not exist on new object', () => {
  expect(newDocument.content.en[1]).toBeUndefined()
  expect(fieldLevelPatches['content.es_ES'][1]).toBeDefined()
  expect(fieldLevelPatches['content.es_ES'][1]._key).toEqual(fieldLevelArticle.content.en[1]._key)
})

test('Arrays will merge objects in the array', () => {
  const documentWithIncompleteObj = getNewFieldLevelDocument()
  const incompleteObj = getNewObject()

  //add a new block with some new content, but not all new content
  documentWithIncompleteObj.content.en.push({
    _key: fieldLevelArticle.content.en[1]._key,
    objectAsField: incompleteObj.objectAsField,
    title: incompleteObj.title,
    //does not include "content" field -- we want that to be merged with the old
  })

  const fieldDocWithMergedObj = BaseDocumentMerger.fieldLevelMerge(
    documentWithIncompleteObj,
    fieldLevelArticle,
    'es_ES',
    'en'
  )

  expect(fieldDocWithMergedObj['content.es_ES'][1].title).toEqual(
    documentWithIncompleteObj.content.en[1].title
  )
  expect(fieldDocWithMergedObj['content.es_ES'][1].objectAsField.title).toEqual(
    documentWithIncompleteObj.content.en[1].objectAsField.title
  )
  //"content" field existed on old doc but not new, so the two coexist happily
  expect(documentWithIncompleteObj.content.en[1].objectAsField.content).toBeUndefined()
  expect(fieldDocWithMergedObj['content.es_ES'][1].objectAsField.content).toBeDefined()
})

test('nested locale fields will be merged', () => {
  const newNestedFields = clone(nestedLanguageFields)
  newNestedFields.pageFields.name.en = 'This is a new page field name'
  newNestedFields.slices[0].en[0].children[0].text = 'This is new slice text'
  const baseDocumentWithNestedFields = {...fieldLevelArticle, ...nestedLanguageFields}
  const newDocumentWithNestedFields = getDeserialized(
    {...fieldLevelArticle, ...newNestedFields},
    'field'
  )
  const nestedFieldLevelPatches = BaseDocumentMerger.fieldLevelMerge(
    newDocumentWithNestedFields,
    baseDocumentWithNestedFields,
    'es_ES',
    'en'
  )

  expect(nestedFieldLevelPatches['slices[0].es_ES'][0].children[0].text).toEqual(
    newDocumentWithNestedFields.slices[0].en[0].children[0].text
  )
  expect(nestedFieldLevelPatches['pageFields.name.es_ES']).toEqual(
    newDocumentWithNestedFields.pageFields.name.en
  )
})
