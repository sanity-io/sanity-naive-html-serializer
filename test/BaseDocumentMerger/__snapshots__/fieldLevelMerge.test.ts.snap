// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global field level snapshot test 1`] = `
Object {
  "_id": "drafts.2947533e-1ea5-4116-955b-339608d3445d",
  "_rev": "l58oha-n06-1s4-f6i-74g6cuakl",
  "_type": "fieldLevelArticle",
  "config.es_ES": Object {
    "_type": "objectField",
    "nestedArrayField": Array [
      Object {
        "_key": "f49b4d7e3e51",
        "_type": "block",
        "children": Array [
          Object {
            "_key": "a6170f21181c",
            "_type": "span",
            "marks": Array [],
            "text": "New text",
          },
        ],
        "markDefs": Array [],
        "style": "normal",
      },
    ],
    "objectAsField": Object {
      "_type": "childObjectField",
      "content": Array [
        Object {
          "_key": "869168fe3a2a",
          "_type": "block",
          "children": Array [
            Object {
              "_key": "10d423df228a",
              "_type": "span",
              "marks": Array [],
              "text": "This is block text 2 levels deep",
            },
          ],
          "markDefs": Array [],
          "style": "normal",
        },
      ],
      "title": "A new nested title",
    },
    "title": "A new title",
  },
  "content.es_ES": Array [
    Object {
      "_key": "e2a39d768ff8",
      "_type": "block",
      "children": Array [
        Object {
          "_key": "randomKey-18",
          "_type": "span",
          "marks": Array [],
          "text": "New block text",
        },
      ],
      "markDefs": Array [],
      "style": "normal",
    },
    Object {
      "_key": "271b0c6ee984",
      "_type": "objectField",
      "objectAsField": Object {
        "_type": "childObjectField",
        "content": Array [
          Object {
            "_key": "4f24f6fbfae7",
            "_type": "block",
            "children": Array [
              Object {
                "_key": "d68f4288f5b7",
                "_type": "span",
                "marks": Array [],
                "text": "This is block text in a nested object in an object in top-level block text.",
              },
            ],
            "markDefs": Array [],
            "style": "normal",
          },
        ],
        "title": "This is a nested object in an object in top-level block text.",
      },
      "title": "This is an object in top-level block text.",
    },
  ],
  "snippet.es_ES": "A new document snippet",
  "tags.es_ES": Array [
    "tag 1",
    "tag 2",
    "tag 3",
  ],
  "title.es_ES": "A new document title",
}
`;
